fix: fetch param files on demand

hanabi1224 opened the pull request:
## Summary of changes

<!-- Please write a comprehensive summary of your changes and what was the motivation behind them -->

Changes introduced in this pull request:

-

## Reference issue to close (if applicable)

<!-- Include the issue reference this pull request is connected to -->
<!-- See more keywords here https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword -->
<!--(e.g. Closes #1)-->

Closes https://github.com/ChainSafe/forest/issues/2843

## Other information and links

<!-- Add any other context about the pull request here. Those might be helpful links based on your investigation, relevant commits from this or other repositories or anything else -->

## Change checklist

<!-- Please add a changelog entry for your change if needed. -->
<!-- Follow this format https://keepachangelog.com/en/1.0.0/ -->

- [ ] I have performed a self-review of my own code,
- [ ] I have made corresponding changes to the documentation,
- [ ] I have added tests that prove my fix is effective or that my feature works (if possible),
- [ ] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

<!-- Thank you 🔥 -->

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

lemmih commented on utils/paramfetch/src/lib.rs:
```diff
@@ -72,6 +75,40 @@ pub fn set_proofs_parameter_cache_dir_env(data_dir: &Path) {
+/// Ensures the parameter files are downloaded to cache dir (async version)
+pub async fn ensure_params_downloaded_async() -> anyhow::Result<()> {
+    lazy_static::lazy_static! {
+        static ref LOCK: tokio::sync::Mutex<()> = tokio::sync::Mutex::new(());
```
If yes, add the reasons as comments in the code.

hanabi1224 replied:
This function is not longer needed

lemmih reviewed (changes requested):
Could you describe your changes in the PR description? This is especially important since your solution differs from the solution proposed in the issue.

hanabi1224 commented:
> Could you describe your changes in the PR description? This is especially important since your solution differs from the solution proposed in the issue.

Updated desc and moved the check to right after snapshot import and before block validation as discussed

hanabi1224 commented:
> This change does not suppress downloading param files, right? It only pushes it for later to speed-up the tests?

Yes

LesnyRumcajs commented:
> > This change does not suppress downloading param files, right? It only pushes it for later to speed-up the tests?
> 
> Yes

I am not 100% convinced that this is the way. I mean, we accommodated the tests, but what is likely more important is to accommodate the overall Forest experience. Did you consider downloading the param files while downloading/importing the snapshot? For a regular Forest user, it would be a startup time improvement.

hanabi1224 commented:
> Let's implement that directly.

Done.  @lemmih  @LesnyRumcajs

LesnyRumcajs commented on utils/genesis/src/lib.rs:
```diff
@@ -114,17 +115,42 @@ where
+    // here because it defaults to `None` in config, making parallelization
+    // disabled by default
+    let ensure_params_downloaded_task_handle =
+        tokio::spawn(forest_utils::proofs_api::paramfetch::ensure_params_downloaded());
```
I am not convinced that `import_chain` is a good place for spawning the params download task. I mean, as a developer, I wouldn't expect such a method to download lots of relatively large files.

Could we somehow refactor this to have those separated? I don't have anything precise in mind. Perhaps splitting the method into smaller ones and a bit of renaming will do.

hanabi1224 replied:
Done

lemmih reviewed (changes requested):
As discussed in Slack, `import_chain` can be refactored to simplify the code significantly.

LesnyRumcajs commented on forest/daemon/src/daemon.rs:
```diff
@@ -538,37 +538,54 @@ async fn prompt_snapshot_or_die(
 async fn sync_from_snapshot<DB>(
     config: &Config,
     state_manager: &Arc<StateManager<DB>>,
+    halt_after_import: bool,
```
This is a cosmetic thing, but this makes it hard to understand for the fellow devs. If you are passing `halt_after_import` boolean, I'd expect this method would do the halting itself, which isn't the case (and rightly so).

`sync_from_snapshot(&config, &state_manager, opts.halt_after_import).`

A bit more expressive form would be

`sync_from_snapshot(&config, &state_manager, ProofParametersFetch::EnsureDownloaded).`

what do you think?

hanabi1224 replied:
Updated

LesnyRumcajs commented on forest/daemon/src/daemon.rs:
```diff
@@ -538,37 +541,48 @@ async fn prompt_snapshot_or_die(
+        let ensure_params_downloaded_task_handle = if proof_params_required {
+            Some(tokio::spawn(async {
+                // Add a small delay to avoid import chain progress bar being interrupted.
+                tokio::time::sleep(Duration::from_secs(1)).await;
```
This is a bit brittle approach, but let's investigate how to sync this properly in another issue, shall we?

lemmih commented on forest/daemon/src/daemon.rs:
```diff
@@ -538,37 +541,48 @@ async fn prompt_snapshot_or_die(
+            ensure_params_downloaded_task_handle.await??;
+        }
+
+        if let Some(validate_height) = validate_height {
+            validate_chain(state_manager, validate_height).await?;
         }
```
The `sync_from_snapshot` shouldn't re-validate the chain. You can hoist it out.

hanabi1224 replied:
As I understand, the whole point of this function is to combine the process of snapshot import and validation and make it cancellable. Would it be better to rename the function to `sync_from_snapshot_and_validate`?

```rust
tokio::select! {
        ret = sync_from_snapshot(&config, &state_manager, proof_params_required).fuse() => {
            if let Err(err) = ret {
                services.shutdown().await;
                return Err(err);
            }
        },
        _ = tokio::signal::ctrl_c() => {
            services.shutdown().await;
            return Ok(db);
        },
        _ = terminate.recv() => {
            services.shutdown().await;
            return Ok(db);
        },
        _ = shutdown_recv.recv() => {
            services.shutdown().await;
            return Ok(db);
        },
    }
```

lemmih replied:
Hmm, the ctrl-c handling is bothering me to no end. It happens so often that ctrl-c is ignored. How about we move the ctrl-c handling out into `main.rs`? That way, the daemon doesn't have to worry about it.

hanabi1224 replied:
@lemmih When I search in the code, I can find 3 places that handle ctrl-c signal in forest daemon, I assume it's duplicated here for a reason.  @elmattic could u share ur insights on whether it is possible to remove the ctrl-c handling from here?

lemmih replied:
There's no good reason to handle interrupts in three places. I'll have a go at refactoring it in a separate PR.

hanabi1224 replied:
> There's no good reason to handle interrupts in three places. I'll have a go at refactoring it in a separate PR.

@lemmih That makes sense, created an issue https://github.com/ChainSafe/forest/issues/2947 and mark this PR as blocked

lemmih replied:
The interrupt code has been merged. Now we can continue to refactor the proof parameter handling.

hanabi1224 replied:
Done

hanabi1224 commented:
Blocked by https://github.com/ChainSafe/forest/issues/2947

lemmih commented on forest/daemon/src/daemon.rs:
```diff
@@ -396,20 +398,43 @@ pub(super) async fn start(
+        Some(0)
+    };
+
+    let proof_params_required = !opts.halt_after_import || validate_height.is_some();
+    let ensure_params_downloaded_task_handle = if proof_params_required {
+        Some(tokio::spawn(async {
+            // Add a small delay to avoid import chain progress bar being interrupted.
+            tokio::time::sleep(Duration::from_millis(100)).await;
+            forest_utils::proofs_api::paramfetch::ensure_params_downloaded().await
+        }))
+    } else {
+        None
+    };
```
I don't think the performance benefits are worth the extra complexity. Let's just ensure that the parameters have been downloaded right before `validate_chain` and right before `p2p_service` is spawned.

hanabi1224 replied:
Fixed.

lemmih commented on forest/daemon/src/daemon.rs:
```diff
@@ -396,20 +398,43 @@ pub(super) async fn start(
+    let validate_height = if config.client.snapshot {
+        config.client.snapshot_height
+    } else {
+        Some(0)
```
Oh, and let's set this to `None`.

hanabi1224 replied:
Fixed.

lemmih commented on forest/daemon/src/daemon.rs:
```diff
@@ -539,31 +553,17 @@ where
 {
     if let Some(path) = &config.client.snapshot_path {
```
Let's inline `sync_from_snapshot`.

hanabi1224 replied:
Done.

lemmih commented on utils/forest_utils/src/proofs_api/paramfetch.rs:
```diff
@@ -72,6 +76,28 @@ pub fn set_proofs_parameter_cache_dir_env(data_dir: &Path) {
 }
 
+/// Ensures the parameter files are downloaded to cache dir
+pub async fn ensure_params_downloaded() -> anyhow::Result<()> {
```
Let's remove `CHECKED` and `LOCK`. We can add them back if/when we decide on downloading proof-parameters in the background.

lemmih commented on forest/daemon/src/daemon.rs:
```diff
@@ -388,19 +389,41 @@ pub(super) async fn start(
+    } else {
+        None
+    };
+    let proof_params_required = !opts.halt_after_import || validate_height.is_some();
+    if proof_params_required {
+        forest_utils::proofs_api::paramfetch::ensure_params_downloaded().await?;
+    }
```
Instead of this logic, you can just call `ensure_params_downloaded` twice. Then the logic is encoded in the control flow.

lemmih commented on forest/daemon/src/daemon.rs:
```diff
@@ -388,19 +389,41 @@ pub(super) async fn start(
+        forest_utils::proofs_api::paramfetch::ensure_params_downloaded().await?;
+    }
+
+    if let Some(validate_height) = validate_height {
+        validate_chain(&state_manager, validate_height).await?;
+    }
```
We can inline `validate_height` and just have two `if` statements here.

hanabi1224 replied:
Done.

lemmih reviewed (approved):
Before:
```
        PASS [  20.027s] forest-daemon::import_snapshot_tests importing_bad_snapshot_should_fail
```
After:
```
        PASS [   3.336s] forest-daemon::import_snapshot_tests importing_bad_snapshot_should_fail
```
